test/cpp/end2end/xds/xds_fallback_end2end_test.cc

```diff
@@ -0,0 +1,317 @@
+  CheckRpcSendOk(DEBUG_LOCATION);
+  EXPECT_EQ(backends_[0]->backend_service()->request_count(), 2);
+  EXPECT_EQ(backends_[1]->backend_service()->request_count(), 0);
+}
+
+TEST_P(XdsFallbackTest, PerAuthorityFallback) {
+  auto fallback_balancer2 = CreateAndStartBalancer("Fallback for Authority2");
+  // Use cleanup in case test assertion fails
+  auto balancer2_cleanup =
+      absl::MakeCleanup([&]() { fallback_balancer2->Shutdown(); });
+  grpc_core::testing::ScopedEnvVar fallback_enabled(
+      "GRPC_EXPERIMENTAL_XDS_FALLBACK", "1");
+  grpc_core::testing::ScopedExperimentalEnvVar env_var(
+      "GRPC_EXPERIMENTAL_XDS_FEDERATION");
+  const char* kAuthority1 = "xds1.example.com";
+  const char* kAuthority2 = "xds2.example.com";
+  constexpr absl::string_view kServer1Name = "server1.example.com";
+  constexpr absl::string_view kServer2Name = "server2.example.com";
+  // Authority1 uses balancer_ and fallback_balancer_
+  // Authority2 uses balancer_ and fallback_balancer2
+  XdsBootstrapBuilder builder;
+  builder.SetServers({balancer_->target()});
+  builder.AddAuthority(kAuthority1,
+                       {balancer_->target(), fallback_balancer_->target()});
+  builder.AddAuthority(kAuthority2,
+                       {balancer_->target(), fallback_balancer2->target()});
+  InitClient(builder);
+  CreateAndStartBackends(4);
+  SetXdsResourcesForServer(fallback_balancer_.get(), 0, kServer1Name,
+                           kAuthority1);
+  SetXdsResourcesForServer(fallback_balancer2.get(), 1, kServer2Name,
+                           kAuthority2);
+  SetXdsResourcesForServer(balancer_.get(), 2, kServer1Name, kAuthority1);
+  SetXdsResourcesForServer(balancer_.get(), 3, kServer2Name, kAuthority2);
+  // Primary balancer is down, using the fallback servers
+  balancer_->ads_service()->ForceADSFailure(
+      Status(StatusCode::RESOURCE_EXHAUSTED, kErrorMessage));
+  // Create second channel to new target URI and send 1 RPC.
+  auto authority1_stub = grpc::testing::EchoTestService::NewStub(CreateChannel(
+      /*failover_timeout_ms=*/0, std::string(kServer1Name).c_str(),
+      kAuthority1));
+  auto authority2_stub = grpc::testing::EchoTestService::NewStub(CreateChannel(
+      /*failover_timeout_ms=*/0, std::string(kServer2Name).c_str(),
+      kAuthority2));
+  ExpectBackendCall(authority1_stub.get(), 0, DEBUG_LOCATION);
+  ExpectBackendCall(authority2_stub.get(), 1, DEBUG_LOCATION);
+  // Primary balancer is up, its data will be used now.
+  balancer_->ads_service()->ClearADSFailure();
+  auto deadline = absl::Now() + absl::Seconds(5);
+  while (absl::Now() < deadline &&
+         (backends_[2]->backend_service()->request_count() == 0 ||
+          backends_[3]->backend_service()->request_count() == 0)) {
+    ClientContext context;
+    EchoRequest request;
+    EchoResponse response;
+    RpcOptions().SetupRpc(&context, &request);
+    Status status = authority1_stub->Echo(&context, request, &response);
+    if (!status.ok()) {
```

Comment:
This should no longer fail now that we're using the same resource names on the primary and fallback servers.